### PR TITLE
refactor: only `build` evm or zksync

### PR DESCRIPTION
**Pull request by Karrq**
## Motivation
As outlined in #335  & #379, not using `--zksync` shouldn't trigger compilation with zksolc, as this can be quite confusing for users.
Additionally, using `--zksync` shouldn't trigger compilation with solc, as some contracts aren't compatible with it (like contracts with era extensions #380)

## Solution
Only `build` with the appropriate compiler.
Future PRs will apply a similar fix for `create` and for `script` subcommands.

### CODE REVIEW

File: crates/forge/bin/cmd/build.rs [hunk 1]
```diff
@@ -87,35 +87,36 @@ impl BuildArgs {
...
+            }
+        } else {
             let zk_compiler = ProjectCompiler::new()
                 .print_names(self.names)
                 .print_sizes(self.sizes)
                 .quiet(self.format_json)
                 .bail(!self.format_json);
-            let zk_output = zk_compiler.zksync_compile(&project)?;
+
+            let output = zk_compiler.zksync_compile(&project)?;
```
Comment (Jrigada) [inline crates/forge/bin/cmd/build.rs:113]: I would maintain zk_output to be clear which one is it faster but given it will only run one branchf of the condition I guess its fine!

**Review by Jrigada — APPROVED** [review]
Great work

File: crates/forge/bin/cmd/create.rs [hunk 2]
```diff
@@ -157,123 +157,135 @@ impl CreateArgs {
...
-                                    .unwrap()
-                                    .to_vec();
-                                visited_bytecodes.insert(fdep_bytecode);
                             }
+
+                            visited_bytecodes.insert(bytecode.clone());
+                            let factory_deps: Vec<Vec<u8>> =
+                                visited_bytecodes.into_iter().collect();
+                            let zk_data = ZkSyncData { bytecode, bytecode_hash, factory_deps };
+                            (abi, bin.object, Some(zk_data))
```
Comment (nbaztec) [inline crates/forge/bin/cmd/create.rs:286]: Can we reduce this nesting somehow? It's a bit hard to follow the logic of option matching

File: crates/forge/bin/cmd/create.rs [hunk 3]
```diff
@@ -157,123 +157,135 @@ impl CreateArgs {
...
-
-                        let factory_dependencies_map =
-                            factory_dependencies.expect("factory deps not found");
-                        let mut visited_paths = HashSet::new();
-                        let mut visited_bytecodes = HashSet::new();
-                        let mut queue = VecDeque::new();
-
-                        for dep in factory_dependencies_map.values() {
-                            queue.push_back(dep.clone());
+                    match (&mut output, &mut zk_output) {
```
Comment (nbaztec) [inline crates/forge/bin/cmd/create.rs:177]: Please move this to an enum, it's cleaner:
```rs
enum CompileMode {
  Evm(output),
  Zk(output),
}
```

File: crates/forge/bin/cmd/create.rs (hunk 2 above)
Comment (Karrq) [reply crates/forge/bin/cmd/create.rs:286]: I think we should re-extract this code into a seprate function (to lookup factory deps) since it does make it a bit longer

File: crates/forge/bin/cmd/create.rs [hunk 4]
```diff
@@ -157,126 +158,128 @@ impl CreateArgs {
...
                     }
+
+                    visited_bytecodes.insert(bytecode.clone());
+                    let factory_deps: Vec<Vec<u8>> = visited_bytecodes.into_iter().collect();
+                    let zk_data = ZkSyncData { bytecode, bytecode_hash, factory_deps };
+                    Ok((abi, bin.object, Some(zk_data)))
                 };
 
+                let (abi, bytecode_object, zk_data) =
+                    output.as_mut().either(process_evm_output, process_zk_output)?;
```
Comment (nbaztec) [inline crates/forge/bin/cmd/create.rs:281]: Can we use `either_into` here? (provided we don't need the output)

**Review by nbaztec — APPROVED** [review]

File: crates/forge/bin/cmd/create.rs (hunk 4 above)
Comment (Karrq) [reply crates/forge/bin/cmd/create.rs:281]: we _do_ need the `output` in later iterations of the loop, and `either_into` calls `.into()` on L or R but we don't have that impl here
